AzureTable/Microsoft.DataTransfer.AzureTable/Sink/TableAPIBulkSinkAdapter.cs

```diff
@@ -0,0 +1,172 @@
+        private BatchSizeTracker batchSizeTracker;
+
+        public int MaxDegreeOfParallelism
+        {
+            get { return 1; }
+        }
+
+        public TableAPIBulkSinkAdapter(string connectionString, string tableName, 
+            bool overwrite, long maxInputBufferSizeInBytes, int throughput, int batchSize)
+        {
+            _connectionString = connectionString;
+            _tableName = tableName;
+            _overwrite = overwrite;
+            _maxInputBufferSizeInBytes = maxInputBufferSizeInBytes;
+            _throughput = throughput;
+            _maxLengthInBytesPerBatch = batchSize;
+
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(_connectionString);
+
+            TableConnectionPolicy connectionPolicy = new TableConnectionPolicy()
+            {
+                UseDirectMode = true,
+                UseTcpProtocol = true,
+            };
+
+            CloudTableClient tableClient = storageAccount.CreateCloudTableClient(connectionPolicy: connectionPolicy);
+            cloudtable = tableClient.GetTableReference(_tableName);
+        }
+
+        public async Task InitializeAsync(CancellationToken cancellation)
+        {
+            inputSizeTracker = new InputSizeTracker(maxLengthInBytesPerDocument, _maxInputBufferSizeInBytes);
+            batchSizeTracker = new BatchSizeTracker(_maxLengthInBytesPerBatch, inputSizeTracker);
+            dict = new ConcurrentDictionary<string, TableBatchOperation>();
+            await cloudtable.CreateIfNotExistsAsync(IndexingMode.Consistent, _throughput, cancellation);
+        }
+
+        public async Task WriteAsync(IDataItem dataItem, CancellationToken cancellation)
+        {
+            var item = GetITableEntityFromIDataItem(dataItem);
+
+            inputSizeTracker.Add(item);
+
+            if (!dict.ContainsKey(item.PartitionKey))
+            {
+                dict[item.PartitionKey] = new TableBatchOperation();
+            }
+
+            var batchOperation = dict[item.PartitionKey];
+            batchOperation.InsertOrReplace(item);
+
+            if (inputSizeTracker.HasExceededInputLimits())
+            {
+                await FlushToTableAsync(cancellation);
+            }
+        }
+
+        public async Task CompleteAsync(CancellationToken cancellation)
+        {
+            if (!inputSizeTracker.HasExceededInputLimits())
+            {
+                await FlushToTableAsync(cancellation);
+            }
+        }
+
+        public void Dispose()
+        {
+        }
+
+        // Flush out all data to the Cosmos DB Table sink.
+        private async Task FlushToTableAsync(CancellationToken cancellation)
+        {
+            try
+            {
+                List<Exception> exceptions = new List<Exception>();
+
+                foreach (var kv in dict)
+                {
+                    var batchOperation = kv.Value;
+                    if (batchOperation.Count > 0)
```

Comment:
Which scenario will have no entries in batch?

Reply:
Its just a safety check. In earlier iterations of the code, I think the code used to split the Tableoperations against a partition key into batches and remove from ks.Value and add to a new list.
And, it so happened that the dictionary "dict" was not cleared out in a finally block. That left kv.Key with a zero length kv.Value list of TableOperations.

We are now clearing out the dictionary "dict" in the finally block... but I left it around as a safety check.

Reply:
Spoke with Kiran. No need to fix.